### ISSUE TITLE
tensor subclass and `__torch_function__` Performance issues

### ISSUE THREAD

**Issue by mattangus** [pytorch/pytorch#73860]
### 🐛 Describe the bug

Sorry If i'm submitting this as the wrong issue type. I would like to subclass `torch.Tensor` to add some additional data/functionality for my use case. This is useful because I would rather have the data encapsulated inside the object rather than having to pass around tuples of tuples of... For example this can be very useful to store meta data, with the advantage of not needing to change existing interfaces in order to accommodate new data being added to the subclass class.

I decided to run some experiments as I was seeing some performance issues. Here is the code to reproduce my experiments. I opted for a bare minimum tensor subclass just to rule out that my code was causing the performance hit. However the actual use case I would like to do is to modify the `__torch_function__` to process `extra_data` with different operations.

```python
from time import sleep, time
import torch
from torchvision.models.mobilenetv3 import mobilenet_v3_large, mobilenet_v3_small
from torchvision.models.efficientnet import efficientnet_b2
from torchvision.models.vgg import vgg16, vgg11
from torchvision.models.alexnet import alexnet
from torchvision.models.googlenet import googlenet

class MyTensor(torch.Tensor):

    extra_data: str = ""

    def __repr__(self):
        return f"extra_data: {self.extra_data}\ndata:\n{super().__repr__()}"

    @classmethod
    def __torch_function__(cls, func, types, args=(), kwargs=None):
        return super().__torch_function__(func, types, args, kwargs)


def do_opt(model, img: torch.Tensor, opt: torch.optim.Optimizer):
    result = model(img)

    opt.zero_grad()
    # just use any loss function
    loss = torch.mean(torch.square(result - 1))
    loss.backward()
    opt.step()

def run_test(model_func, n=100, burn_in=5):
    print(f"testing model {model_func.__name__}")

    model = model_func().cuda()

    img = torch.rand([32, 3, 256, 256]).cuda()

    opt = torch.optim.Adam(model.parameters())

    for i in range(burn_in):
        do_opt(model, img, opt)

    start1 = time()
    for i in range(n):
        do_opt(model, img, opt)
    end1 = time()

    img = img.as_subclass(MyTensor)
    img.extra_data = "test"

    start2 = time()
    for i in range(n):
        do_opt(model, img, opt)
    end2 = time()

    tensor_time = 1000 * (end1 - start1) / n
    mytensor_tme = 1000 * (end2 - start2) / n
    pct = 100 * (mytensor_tme - tensor_time) / tensor_time
    sign = "+" if pct >= 0 else ""

    print(f"torch.Tensor: {tensor_time:0.2f}ms per iteration")
    print(f"    MyTensor: {mytensor_tme:0.2f}ms per iteration")
    print(f"% difference: {sign}{pct:0.2f}%")

def main():

    models_to_test = [
        mobilenet_v3_large,
        mobilenet_v3_small,
        efficientnet_b2,
        vgg16,
        vgg11,
        alexnet
    ]

    for model_func in models_to_test:
        run_test(model_func)

if __name__ == "__main__":
    main()
```

Here is the output with pytorch nightly.

```
testing model mobilenet_v3_large
torch.Tensor: 54.04ms per iteration
    MyTensor: 66.64ms per iteration
% difference: +23.31%
testing model mobilenet_v3_small
torch.Tensor: 25.54ms per iteration
    MyTensor: 31.23ms per iteration
% difference: +22.25%
testing model efficientnet_b2
torch.Tensor: 144.74ms per iteration
    MyTensor: 169.72ms per iteration
% difference: +17.26%
testing model vgg16
torch.Tensor: 286.61ms per iteration
    MyTensor: 324.19ms per iteration
% difference: +13.11%
testing model vgg11
torch.Tensor: 161.03ms per iteration
    MyTensor: 181.56ms per iteration
% difference: +12.75%
testing model alexnet
torch.Tensor: 26.12ms per iteration
    MyTensor: 27.51ms per iteration
% difference: +5.29%
```
Here is the output with pytorch stable:
```
testing model mobilenet_v3_large
torch.Tensor: 55.69ms per iteration
    MyTensor: 67.54ms per iteration
% difference: +21.27%
testing model mobilenet_v3_small
torch.Tensor: 25.23ms per iteration
    MyTensor: 29.33ms per iteration
% difference: +16.26%
testing model efficientnet_b2
torch.Tensor: 142.44ms per iteration
    MyTensor: 169.52ms per iteration
% difference: +19.01%
testing model vgg16
torch.Tensor: 291.71ms per iteration
    MyTensor: 330.15ms per iteration
% difference: +13.18%
testing model vgg11
torch.Tensor: 163.73ms per iteration
    MyTensor: 185.06ms per iteration
% difference: +13.03%
testing model alexnet
torch.Tensor: 26.09ms per iteration
    MyTensor: 27.48ms per iteration
% difference: +5.33%
```

As you can see in some cases the 

### Versions

I ran these tests with two versions of pytorch.

Nightly:

```
Collecting environment information...
PyTorch version: 1.12.0.dev20220307+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-6ubuntu2) 7.5.0
Clang version: Could not collect
CMake version: version 3.18.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, Nov 26 2021, 20:14:08)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-100-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.0.194
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 460.73.01
cuDNN version: Probably one of the following:
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] efficientnet-pytorch==0.7.1
[pip3] mypy==0.800
[pip3] mypy-extensions==0.4.3
[pip3] mypy-protobuf==1.23
[pip3] numpy==1.21.3
[pip3] pytorch-lightning==1.4.4
[pip3] torch==1.12.0.dev20220307+cu113
[pip3] torchaudio==0.12.0.dev20220307+cu113
[pip3] torchmetrics==0.5.0
[pip3] torchvision==0.13.0.dev20220307+cu113
[conda] Could not collect
```

and stable:
```
Collecting environment information...
PyTorch version: 1.10.2+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-6ubuntu2) 7.5.0
Clang version: Could not collect
CMake version: version 3.18.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, Nov 26 2021, 20:14:08)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-100-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.0.194
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 460.73.01
cuDNN version: Probably one of the following:
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.4
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] efficientnet-pytorch==0.7.1
[pip3] mypy==0.800
[pip3] mypy-extensions==0.4.3
[pip3] mypy-protobuf==1.23
[pip3] numpy==1.21.3
[pip3] pytorch-lightning==1.4.4
[pip3] torch==1.10.2+cu113
[pip3] torchaudio==0.10.2+cu113
[pip3] torchmetrics==0.5.0
[pip3] torchvision==0.11.3+cu113
[conda] Could not collect
```

**Comment by rgommers** [comment]
This is quite surprising. Even the absolute numbers - the overhead you show is ~10 ms over 100 iterations, so 100 us per iteration. The overhead for a single `__torch_function__` call should be sub-microsecond.

It would to use `line_profiler`, `pyspy` or similar to get a better sense of where the time is being spent.

**Comment by mattangus** [comment]
Also @rgommers the calculation of time is per iteration. So the added overhead is ~10ms per iteration not 100us.

```python
tensor_time = 1000 * (end1 - start1) / n # <-- divide by number of iters 
mytensor_tme = 1000 * (end2 - start2) / n
```

**Comment by mattangus** [comment]
Also here is the `line_profiler` on my own code.

```
Timer unit: 1e-06 s

Total time: 2.17898 s
File: standalone.py
Function: __torch_function__ at line 16

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    16                                               @classmethod
    17                                               @profile
    18                                               def __torch_function__(cls, func, types, args=(), kwargs=None):
    19     23700    2178980.0     91.9    100.0          return super().__torch_function__(func, types, args, kwargs)

Total time: 13.6049 s
File: standalone.py
Function: do_opt at line 21

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    21                                           @profile
    22                                           def do_opt(model, img: torch.Tensor, opt: torch.optim.Optimizer):
    23       205    2169296.0  10581.9     15.9      result = model(img)
    24                                           
    25       205     229802.0   1121.0      1.7      opt.zero_grad()
    26                                               # just use any loss function
    27       205      13473.0     65.7      0.1      loss = torch.mean(torch.square(result - 1))
    28       205    2610369.0  12733.5     19.2      loss.backward()
    29       205    8581943.0  41863.1     63.1      opt.step()

Total time: 15.8932 s
File: standalone.py
Function: run_test at line 31

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    31                                           @profile
    32                                           def run_test(model_func, n=100, burn_in=5):
    33         1         14.0     14.0      0.0      print(f"testing model {model_func.__name__}")
    34                                           
    35         1    2237087.0 2237087.0     14.1      model = model_func().cuda()
    36                                           
    37         1      28286.0  28286.0      0.2      img = torch.rand([32, 3, 256, 256]).cuda()
    38                                           
    39         1       1838.0   1838.0      0.0      opt = torch.optim.Adam(model.parameters())
    40                                           
    41                                           
    42                                           
    43         6         11.0      1.8      0.0      for i in range(burn_in):
    44         5     386925.0  77385.0      2.4          do_opt(model, img, opt)
    45                                           
    46         1          2.0      2.0      0.0      start1 = time()
    47       101        124.0      1.2      0.0      for i in range(n):
    48       100    5987965.0  59879.7     37.7          do_opt(model, img, opt)
    49         1          2.0      2.0      0.0      end1 = time()
    50                                           
    51         1         16.0     16.0      0.0      img = img.as_subclass(MyTensor)
    52         1          1.0      1.0      0.0      img.extra_data = "test"
    53                                           
    54         1          0.0      0.0      0.0      start2 = time()
    55       101        122.0      1.2      0.0      for i in range(n):
    56       100    7250814.0  72508.1     45.6          do_opt(model, img, opt)
    57         1          2.0      2.0      0.0      end2 = time()
    58                                           
    59         1          1.0      1.0      0.0      tensor_time = 1000 * (end1 - start1) / n
    60         1          1.0      1.0      0.0      mytensor_tme = 1000 * (end2 - start2) / n
    61         1          0.0      0.0      0.0      pct = 100 * (mytensor_tme - tensor_time) / tensor_time
    62                                               
    63         1          0.0      0.0      0.0      sign = "+" if pct >= 0 else ""
    64                                               
    65         1         20.0     20.0      0.0      print(f"torch.Tensor: {tensor_time:0.2f}ms per iteration")
    66         1          3.0      3.0      0.0      print(f"    MyTensor: {mytensor_tme:0.2f}ms per iteration")
    67         1          5.0      5.0      0.0      print(f"% difference: {sign}{pct:0.2f}%")

Total time: 15.8944 s
File: standalone.py
Function: main at line 70

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    70                                           @profile
    71                                           def main():
    72                                           
    73         1          0.0      0.0      0.0      models_to_test = [
    74         1          1.0      1.0      0.0          mobilenet_v3_large,
    75                                                   # mobilenet_v3_small,
    76                                                   # efficientnet_b2,
    77                                                   # vgg16,
    78                                                   # vgg11,
    79                                                   # alexnet
    80                                               ]
    81                                           
    82         2          0.0      0.0      0.0      for model_func in models_to_test:
    83         1   15894375.0 15894375.0    100.0          run_test(model_func)
```

I also stumbled across the [benchmarks folder](https://github.com/pytorch/pytorch/tree/master/benchmarks/overrides_benchmark) and ran the `bench.py` with the following output:

```
Type tensor had a minimum time of 1.485586166381836 us and a standard deviation of 0.04454116424312815 us.
Type SubTensor had a minimum time of 5.297183990478516 us and a standard deviation of 0.25198794901371 us.
Type WithTorchFunction had a minimum time of 3.160715103149414 us and a standard deviation of 0.14244357589632273 us.
Type SubWithTorchFunction had a minimum time of 5.589008331298828 us and a standard deviation of 0.14299638860393316 us.
```

My `__torch_function__` is called 23700 times. So the difference of `~3.81us*23700 = ~90.3ms` for the whole program so ~9ms per iteration. This is pretty close ~10ms difference I am seeing. Does that mean the overhead I'm seeing is to be expected?

**Comment by rgommers** [comment]
> Small correction, this overhead is for the case where `__torch_function__` is basically disabled (i.e. for `Tensor` and `Parameter` by default).
> 
> It is perfectly normal to have a ~10 ms overhead depending on what the function is doing.

@hameerabbasi thanks for the correction. The code in the issue description is:
```python
    @classmethod
    def __torch_function__(cls, func, types, args=(), kwargs=None):
        return super().__torch_function__(func, types, args, kwargs)
```
where `super()` is `torch.Tensor`. So the code is doing nothing specific yet. The only extra work is in `_convert`:
```python
def _convert(ret, cls):
    if cls is Tensor:
        return ret

    if isinstance(ret, Tensor) and not isinstance(ret, cls):
        ret = ret.as_subclass(cls)

    if isinstance(ret, (tuple, list)):
        # Also handles things like namedtuples
        ret = type(ret)(_convert(r, cls) for r in ret)

    return ret
```



> My `__torch_function__` is called 23700 times. So the difference of `~3.81us*23700 = ~90.3ms` for the whole program so ~9ms per iteration. This is pretty close ~10ms difference I am seeing. Does that mean the overhead I'm seeing is to be expected?

That indeed looks like a consistent story now. But looking at the benchmarks, it's the fact that you're subclassing `Tensor` mostly, rather than adding a custom `__torch_function__` to it. The difference in the benchmarks between `SubTensor` and `SubWithTorchFunction` is a lot smaller than that between `Tensor` and `SubTensor`.

You're seeing >100 calls to `__torch_function__` per `do_opt` step. Depending on your use case with `extra_data`, could you perhaps convert to `Tensor` to avoid those calls, which look like they're coming from `loss.backward()` and/or `opt.step()` mostly?
